doc/developer/design/20231027_refresh_mvs.md

```diff
@@ -0,0 +1,279 @@
+
+For showing cluster schedules, I'm thinking of doing a similar thing as in `mz_materialized_view_refresh_options`, because eventually the `SCHEDULE =` option will allow for specifying multiple schedules. So I'd create a new table in `mz_internal` called `mz_cluster_schedules`, and it would have a row for each non-trivial (non-`MANUAL`) schedule option of each cluster. This would currently be only `SCHEDULE = ON REFRESH`. Columns would be `(cluster_id text, schedule_type text, refresh_rehydration_time_estimate interval)`. In the `schedule_type`, we would currently always have "ON REFRESH". (Eventually, we'll probably also want a `next_scheduled_turn_on`, but this doesn't seem so urgent. It will get more important when we'll be choosing the warmup time automatically.)
+
+For seeing whether a cluster is currently turned on, the user can simply look at `mz_cluster_replicas`, because we currently turn clusters On/Off by just creating/dropping replicas. We might also add a builtin view for showing this information in a more focused way.
+
+For the automatic cluster scheduling history, the user can look at `mz_audit_events`. This has a `details` column, which is a JSON blob, where I'm planning to add the `reason` for turning on a cluster, i.e., which materialized views were in need of a refresh. (See Nikhil's comment [here](https://github.com/MaterializeInc/materialize/pull/26401#pullrequestreview-1981986544).) There is also the `mz_cluster_replica_history` view, which takes its info from `mz_audit_events`, and presents the info in a nicer form. I could add a new reason column to this view.
+
+We'll also want to show rehydration times from the last several refreshes, to help users set the `REHYDRATION TIME ESTIMATE` of clusters. I'm thinking to create a new table `mz_internal.mz_compute_hydration_history (replica_id text, rehydration_time interval)`, which would have one row for each replica creation, and it would show the time it took to rehydrate the replica when it was created. (The user can join this with `mz_cluster_replica_history` to know which cluster the replica belonged to, replica size, etc.) Btw. this doesn't need to be constrained to clusters involving `REFRESH` MVs; this info seems generally useful for any compute cluster. If we want to make it even more useful generally, we might want to add one row not just for each replica creation, but also each replica restart, so that we'll show the rehydrations that happen at system upgrade restarts. In this case, we'll probably need also a `time` column, and then `(replica_id, time)` would be a composite key. For this general version, we might have to truncate the relation to keep it from growing too big.
```

Comment:
I like the general version (i.e., the version that records rehydration time) myself! I think you'll need to handle truncation in both versions, as otherwise, even in the only-on-creation version, running `CREATE REPLICA; .. wait for hydration; DROP REPLICA` in a hot loop would cause `mz_compute_hydration_history` to grow without bound.

Reply:
One thing I'm planning to do is have `mz_compute_hydration_statuses` (eventually `mz_hydration_statuses` too, provided we get enough information from storage) report a hydration time as well. This also gives us a rough per-replica hydration time through `SELECT replica_id, max(hydration_time) FROM mz_internal.mz_compute_hydration_statuses GROUP BY replica_id`. I see advantages and disadvantages to the proposed solution:

* No history. This is a clear disadvantage.
* Not limited to dataflows that were installed at replica creation. This can be both an advantage and a disadvantage. If the set of installed dataflows wasn't changed since the last hydration, `mz_compute_hydration_history` would give a more accurate time. But if dataflows were created or dropped since then, the time in the history might be way off from the time actually required to do the next hydration, while `mz_compute_hydration_statuses` would still provide a good ballpark estimate.

You can get a history from `mz_compute_hydration_statuses` by increasing the retention window. And then using `SUBSCRIBE AS OF` like the console does for other compute introspection collections. It's not a great solution as (a) it doesn't prevent the issue Nikhil pointed out, (b) we'd maybe have to remove transient collections from that table, to keep its size manageable, and (c) usability is bad in SQL.

So I think if we need a history, having a separate table is the way. One thing we should clarify though is what it means for a replica to hydrate. Which dataflows have to be hydrated before a replica is considered hydrated? Only the ones installed when the replica was started? Or also ones that have been installed during replica hydration?